pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageIdImpl.java

```diff
@@ -63,17 +78,17 @@ public int getPartitionIndex() {
 
     @Override
     public int hashCode() {
-        return (int) (31 * (ledgerId + 31 * entryId) + partitionIndex);
+        return Objects.hash(ledgerId, entryId, partitionIndex, batchIndex);
```

Comment:
This call will involve several allocations on the heap, because `Objects.hash()` takes a `Object[]` and all the longs also need to get boxes into `Long` instances.

Reply:
It seems like `Objects.hash` is used elsewhere in the codebase and I generally prefer it to manually implementing `hashCode`, but I don't have a strong opinion. Happy to change it back.